conda_smithy/configure_feedstock.py

```diff
@@ -421,6 +430,58 @@ def _get_used_key_values_by_input_order(
     return used_key_values
 
 
+def _merge_deployment_target(container_of_dicts, has_macdt):
+    """
+    For a collection of variant dictionaries, merge deployment target specs.
+
+    - The "old" way is MACOSX_DEPLOYMENT_TARGET, the new way is c_stdlib_version;
+      For now, take the maximum to populate both.
+    - In any case, populate MACOSX_DEPLOYMENT_TARGET, as that is the key picked
+      up by https://github.com/conda-forge/conda-forge-ci-setup-feedstock
+    """
+    result = []
+    for var_dict in container_of_dicts:
+        # cases where no updates are necessary
+        if not var_dict.get("target_platform", "dummy").startswith("osx"):
+            result.append(var_dict)
+            continue
+        if "c_stdlib_version" not in var_dict:
+            result.append(var_dict)
+            continue
+        # case where we need to do processing
+        v_stdlib = var_dict["c_stdlib_version"]
+        macdt = var_dict.get("MACOSX_DEPLOYMENT_TARGET", v_stdlib)
+        # error out if someone puts in a range of versions; we need a single version
+        try:
+            cond_update = VersionOrder(v_stdlib) < VersionOrder(macdt)
+        except InvalidVersionSpec:
+            raise ValueError(
+                "both and c_stdlib_version/MACOSX_DEPLOYMENT_TARGET need to be a "
+                "single version, not a version range!"
+            )
+        if v_stdlib != macdt:
+            # determine maximum version and use it to populate both
+            v_stdlib = macdt if cond_update else v_stdlib
+            msg = (
+                "Conflicting specification for minimum macOS deployment target!\n"
+                "If your conda_build_config.yaml sets `MACOSX_DEPLOYMENT_TARGET`, "
+                "please change the name of that key to `c_stdlib_version`!\n"
+                f"Using {v_stdlib}=max(c_stdlib_version, MACOSX_DEPLOYMENT_TARGET)."
+            )
+            # we don't want to warn for recipes that do not use MACOSX_DEPLOYMENT_TARGET
+            # in the local CBC, but only inherit it from the global pinning
+            if has_macdt:
+                warn_once(msg)
+
+        # we set MACOSX_DEPLOYMENT_TARGET to match c_stdlib_version,
+        # for ease of use in conda-forge-ci-setup
+        var_dict["c_stdlib_version"] = v_stdlib
+        var_dict["MACOSX_DEPLOYMENT_TARGET"] = v_stdlib
```

Comment:
This modifies the input dictionary which is undesirable for the `list_of_metas[0].config.input_variants` case.

Reply:
Hm yeah. I really don't like these implicit in-place updates (kinda wrote it like that previously without thinking too much apparently), and so using the explicit `result` was an intentional move away from that. But I still get bitten by it because the in-place updates of `var_dict` haven't gone away.

WDYT about using a `copy.deepcopy` of `container_of_dicts` at the beginning of the loop?

Reply:
> WDYT about using a `copy.deepcopy` of `container_of_dicts` at the beginning of the loop?

IMO, `deepcopy` is a unnecessary blunt hammer in this case.
Something like `var_dict = {**var_dict, "c_stdlib_version": v_stdlib, "MACOSX_DEPLOYMENT_TARGET": v_stdlib}` would suffice, wouldn't it?

Reply:
These copies are what slows down conda build a ton when rerendering. Can we move the copy outside the function and only call it on just on the input variants bits?

Reply:
OK, message received! Thanks both for the input! Should be fixed in https://github.com/conda-forge/conda-smithy/pull/1889/commits/9f4634bc1e15427b9d7dbedfdc6724b573b98a9d